lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->commit_class_changes();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+		if ( false === $this->scanner->scan( $this->descriptor ) ) {
+			return false;
+		}
+
+		return $this;
+	}
+
+
+	public function add_attribute( $name, $value ) {
+		$tag_name = $this->scanner->tag_name;
+		$insert_at = $tag_name->start + $tag_name->length;
+
+		$this->modifications->replacements[] = array(
+			$insert_at,
+			$insert_at,
+			' ' . self::serialize_attribute( $name, $value )
+		);
+
+		return $this;
+	}
+
+
+	public function add_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = true;
+	}
+
+
+	public function set_attribute( $name, $value ) {
+		if ( ! array_key_exists( WP_Tag_Find_Descriptor::comparable( $name ), $this->scanner->attributes ) ) {
+			return $this->add_attribute( $name, $value );
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			self::serialize_attribute( $name, $value )
+		);
+	}
+
+
+	public function remove_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = false;
+	}
+
+
+	public function remove_attribute( $name ) {
+		if ( ! array_key_exists( $name, $this->scanner->attributes ) ) {
+			return $this;
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			''
+		);
+	}
+
+
+	public static function serialize_attribute( $name, $value ) {
+		$value = str_replace( '"', "&quot;", $value );
+
+		return "{$name}=\"{$value}\"";
+	}
+
+
+	public function commit_class_changes() {
+		if ( empty( $this->modifications->class_changes ) ) {
+			return;
+		}
+
+		$existing_class = array_key_exists( 'class', $this->scanner->attributes )
+			? substr(
+				$this->scanner->document,
+				$this->scanner->attributes['class']->value->start,
+				$this->scanner->attributes['class']->value->length
+			)
+			: '';
```

Comment:
I went for storing values in https://github.com/WordPress/gutenberg/pull/42485, but that's only useful in case of class names so storing just indices sounds better 👍

Reply:
I want to get rid of the value token and instead just store start and end of the attribute value itself as well as the location where that's found (including the quotes)

Reply:
and I've considered that for _all_ tokens as well to avoid allocating more objects. I'm still puzzled by why we're getting _any_ out of memory exceptions when running on big objects

Reply:
Weirdly, https://github.com/WordPress/gutenberg/pull/42485 uses less memory despite doing pretty much the same thing as this PR. Perhaps the memory errors here are due to some very specific weird quirk? If you can share reproduction steps, I'll see if I can pinpoint it.

Reply:
for historical reference: the memory use came from the fact that I was storing modifications as a three-array of `[int, int, string]`. when I converted to a class with `start`, `end`, and `text` _properties_ the memory use dropped to 5% of what it was with the array.